design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,599 @@
+  functions:
+  - name: my-cool-function
+    type: Container
+    # Configuration specific to `type: Container` functions.
+    container:
+      # The OCI image to pull and run.
+      image: xkpg.io/my-cool-function:0.1.0
+      # Whether to pull the function Never, Always, or IfNotPresent.
+      imagePullPolicy: IfNotPresent
+      # Secrets used to pull from a private registry.
+      imagePullSecrets:
+      - namespace: crossplane-system
+        name: my-xpkg-io-creds
+      # Note that only resource limits are supported - not requests.
+      # The function will be run with the specified resource limits.
+      resources:
+        limits:
+          memory: 64Mi
+          cpu: 250m
+      # Defaults to 'Isolated' - i.e an isolated network namespace.
+      network: Accessible
+      # How long the function may run before it's killed. Defaults to 10s.
+      timeout: 30s
+    # An x-kubernetes-embedded-resource RawExtension (i.e. an unschemafied)
+    # Kubernetes resource. Passed to the function as the functionConfig block of
+    # its ResourceList.
+    config:
+      apiVersion: database.example.org/v1alpha1
+      kind: Config
+      metadata:
+        name: cloudsql
+      spec:
+        version: POSTGRES_9_6
+```
+
+### Function API
+
+This document proposes that each function uses a `ResourceList` type derived
+from the [KRM function specification][krm-fn-spec] as its input and output.
+
+```yaml
+apiVersion: fn.apiextensions.crossplane.io/v1
+kind: ResourceList
+functionConfig:
+  apiVersion: database.example.org/v1alpha1
+  kind: Config
+  metadata:
+    name: cloudsql
+  spec:
+    version: POSTGRES_9_6
+items:
+- apiVersion: database.example.org/v1alpha1
+  kind: XPostgreSQLInstance
+  metadata:
+    name: my-db
+    annotations:
+      fn.apiextensions.crossplane.io/type: "CompositeResource"
+  spec:
+    parameters:
+      storageGB: 20
+    compositionSelector:
+      matchLabels:
+        provider: gcp
+    writeConnectionSecretToRef:
+      name: db-conn
+```
+
+Like a KRM function, a Composition function uses a `ResourceList` that consists
+of the following top-level fields:
+
+* The `apiVersion` and `kind` (required).
+* A `functionConfig` (optional). This is a [Kubernetes resource][rawextension]
+  with an arbitrary schema that may be used to provide additional configuration
+  to a function. For example a `render-helm-chart` function might use its
+  `functionConfig` to specify which Helm chart to render.
+* An `items` array (required). The XR and zero or more composed resources.
+* A `results` object (optional). Used to communicate information about the
+  result of a function, including warnings and errors.
+
+Each function in the array would be executed as a pipeline, with each function:
+
+1. Taking the XR and zero or more composed resources as input.
+1. Producing the optionally mutated XR, optionally mutated composed resources,
+   and newly rendered composed resources as output.
```

Comment:
**question**: Are there any assumptions for these newly composed resources to adhere to XRM? More specifically, can my function render arbitrary, namespace-scoped resources?

Reply:
They're just rendering the resources that Crossplane should reconcile, so they're under the same restrictions as a contemporary Compositon - rendering XRM resources. To be fair I don't think this is actually enforced in either case (resources array style Composition or functions array style), but it's unsupported and the behavior is undefined.